### ISSUE TITLE
Support for jellying types

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2950]
|<img alt="james@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| james@... reported|
|-|-|
|Trac ID|trac#2950|
|Type|enhancement|
|Created|2007-12-12 01:59:07Z|
|Branch|https://github.com/twisted/twisted/tree/type-jelly-2950-2|

Fixes support for the jellying of type objects and adds unit tests.

I removed the isClassAllowed() test, under the following rationale:
  * isClassAllowed() checks the list of allowed class _instances_ (see allowInstancesOf())
  * I couldn't conceive of a reasonable exploit, since types are inert until instantiated.

Attachments:

* [type_jelly.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2f1/2f180a13eaaa0726017461d6a5ed7b5a8b983bf2/4b1554993f55dd84816f29ebffbcf90b494a7c53.patch) (1516 bytes) - added by james@... on 2007-12-12 02:00:07Z - patch for jelly.py and test_jelly.py

<details><summary>Searchable metadata</summary>

```
trac-id__2950 2950
type__enhancement enhancement
reporter__james____ james@...
priority__normal normal
milestone__ 
branch__branches_type_jelly_2950_2 branches/type-jelly-2950-2
branch_author__james______therve james@..., therve
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1197424747000000 1197424747000000
changetime__1240507641000000 1240507641000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22578]) Branching to 'type-jelly-2950'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22579]) Do it.

Refs [#2950](https://github.com/twisted/twisted/issues/2950)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

I think this is ready to review. I didn't change anything security-related: that looked ok to me. I just handled the `TypeType` case, and added some tests.

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bigdog set owner to bigdog<br>bigdog set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|bigdog set owner to @therve<br>bigdog set status to `new`|
|-|-|

Verified fix matches scope of ticket

Tests cover change and doc strings present

ran on windows and linux with no ill effects

code is ready to go

**Comment by twisted-trac** [comment]
|<img alt="james@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|james@... commented|
|-|-|

The tests supplied with this patch will run with taster = DummySecurityOptions()

They will fail if run against the default globalSecurity object.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

[#3354](https://github.com/twisted/twisted/issues/3354) has been closed as a duplicate of this one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26749]) trivial style updates; copyright date update

refs [#2950](https://github.com/twisted/twisted/issues/2950)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26750]) Branching to 'type-jelly-2950-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [26752]) Merge type-jelly-2950-2

Author: james@..., therve
Reviewer: bigdog, exarkun
Fixes: [#2950](https://github.com/twisted/twisted/issues/2950)

Change jelly to consider instances of `type` to be classes, as it already
considered instances of `ClassType` to be classes.  This causes it to
support unjellying new-style classes in addition to classic classes.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
